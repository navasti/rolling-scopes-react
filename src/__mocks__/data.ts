import { ErrorMessages } from 'appConstants';
import {
  PokemonMoveDetails,
  PokemonTypeDetails,
  PokemonDetails,
  CustomPokemon,
  ErrorsObject,
  PokemonData,
  PokemonMove,
  PokemonType,
  Details,
  Pokemon,
  Lengths,
} from 'types';

export const visible = true;
export const invisible = false;
export const isLoading = false;

export const links = ['Home', 'About', 'Form', 'Invalid'];

export const testingComponentName = 'Test';
export const testingLocation = '/test';

export const testingContent = 'Content';
export const testingTitle = 'Title';

export const pokemonType: PokemonType = {
  name: 'typex',
  url: 'url',
};

export const pokemon: Pokemon = {
  name: 'pokemonix',
  url: 'url',
};

export const pokemonMove: PokemonMove = {
  name: 'movix',
  url: 'url',
};

export const typeModalContentProperties = [
  'Double damage from',
  'Double damage to',
  'Half damage from',
  'Half damage to',
  'No damage from',
  'No damage to',
  'Pokemons',
  'Moves',
] as const;

export const moveModalContentProperties = [
  'Learned by',
  'Priority',
  'Accuracy',
  'Power',
  'Type',
  'PP',
] as const;

export const pokemonModalContentProperties = [
  'Pokemon statistics',
  'Pokemon abilities',
  'Base experience',
  'Height',
  'Weight',
  'Types',
] as const;

export const pokemonData: PokemonData = {
  results: [{ name: 'testName', url: 'testURL' }],
  previous: null,
  next: null,
  count: 1,
};

export const testEmptyErrors: ErrorsObject = {
  birthday: null,
  consent: null,
  gender: null,
  name: null,
  type: null,
};

export const testErrors: ErrorsObject = {
  birthday: ErrorMessages.birthday,
  consent: ErrorMessages.consent,
  gender: ErrorMessages.gender,
  name: ErrorMessages.name,
  type: ErrorMessages.type,
};

export const customPokemon: CustomPokemon = {
  birthday: '01-01-1999',
  name: 'pokemonix',
  gender: 'male',
  avatar: null,
  type: 'fire',
  shiny: true,
};

//

export const lengths: Lengths = { pokemons: 10, moves: 15, types: 20 };

const samplePokemon = (): PokemonDetails => ({
  base_experience: 62,
  name: 'charmander',
  weight: 85,
  height: 6,
  id: 4,
  abilities: [
    {
      ability: { name: 'blaze', url: 'https://pokeapi.co/api/v2/ability/66/' },
      is_hidden: false,
      slot: 1,
    },
  ],
  sprites: {
    front_default: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/4.png',
    back_default:
      'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/back/4.png',
  },
  stats: [
    {
      base_stat: 52,
      effort: 0,
      stat: {
        name: 'attack',
        url: 'https://pokeapi.co/api/v2/stat/2/',
      },
    },
  ],
  types: [
    {
      slot: 1,
      type: {
        name: 'fire',
        url: 'https://pokeapi.co/api/v2/type/10/',
      },
    },
  ],
});

const sampleMove = (): PokemonMoveDetails => ({
  accuracy: 100,
  name: 'tackle',
  priority: 0,
  power: 40,
  id: 33,
  pp: 35,
  learned_by_pokemon: [
    {
      name: 'bulbasaur',
      url: 'https://pokeapi.co/api/v2/pokemon/1/',
    },
  ],
  type: {
    name: 'normal',
    url: 'https://pokeapi.co/api/v2/type/1/',
  },
});

const sampleType = (): PokemonTypeDetails => ({
  moves: [{ name: 'gust', url: 'https://pokeapi.co/api/v2/move/16/' }],
  name: 'flying',
  id: 3,
  damage_relations: {
    double_damage_from: [
      { name: 'electric', url: 'https://pokeapi.co/api/v2/type/13/' },
      { name: 'rock', url: 'https://pokeapi.co/api/v2/type/6/' },
    ],
    double_damage_to: [{ name: 'fighting', url: 'https://pokeapi.co/api/v2/type/2/' }],
    half_damage_from: [{ name: 'bug', url: 'https://pokeapi.co/api/v2/type/7/' }],
    half_damage_to: [{ name: 'steel', url: 'https://pokeapi.co/api/v2/type/9/' }],
    no_damage_from: [{ name: 'ground', url: 'https://pokeapi.co/api/v2/type/5/' }],
    no_damage_to: [],
  },
  pokemon: [
    { pokemon: { name: 'charizard', url: 'https://pokeapi.co/api/v2/pokemon/6/' }, slot: 2 },
  ],
});

const prepareBuilder = <T extends Details>(sample: () => T) => {
  const builder = (object = sample()) => ({
    withName: (name: T['name']) => builder({ ...object, name }),
    withId: (id: T['id']) => builder({ ...object, id }),
    valueOf: () => object,
  });
  return builder;
};

const pokemonBuilder = prepareBuilder(samplePokemon);
const moveBuilder = prepareBuilder(sampleMove);
const typeBuilder = prepareBuilder(sampleType);

export const pokemonsMock: Array<PokemonDetails> = [
  pokemonBuilder().withId(1).withName('bulbasaur'),
  pokemonBuilder().withId(25).withName('pikachu'),
  pokemonBuilder(),
].map((builder) => builder.valueOf());

export const movesMock: Array<PokemonMoveDetails> = [
  moveBuilder().withId(55).withName('water-gun'),
  moveBuilder().withId(50).withName('disable'),
  moveBuilder(),
].map((builder) => builder.valueOf());

export const typesMock: Array<PokemonTypeDetails> = [
  typeBuilder().withId(2).withName('fighting'),
  typeBuilder().withId(5).withName('ground'),
  typeBuilder(),
].map((builder) => builder.valueOf());
