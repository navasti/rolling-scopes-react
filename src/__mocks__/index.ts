export { testOnChange, handleCloseModal } from './handlers';
export { localStorageMock } from './localStorage';

export {
  fetchBaseMock,
  fetchDetailsMock,
  handleResponseMock,
  fetchByParameterMock,
  handleMappedResponseMock,
} from './api';

export {
  links,
  pokemon,
  visible,
  invisible,
  isLoading,
  movesMock,
  typesMock,
  testErrors,
  pokemonData,
  pokemonMove,
  pokemonType,
  testingTitle,
  pokemonsMock,
  customPokemon,
  testingContent,
  testingLocation,
  testEmptyErrors,
  testingComponentName,
  moveModalContentProperties,
  typeModalContentProperties,
  pokemonModalContentProperties,
} from './data';

export {
  modalRef,
  errorRef,
  inputRef,
  selectRef,
  ModalTitle,
  ModalContent,
  TestingElement,
  TestingErrorElement,
} from './elements';
