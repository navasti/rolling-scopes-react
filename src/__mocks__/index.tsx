import { ErrorMessages } from 'appConstants';
import { createRef, RefObject } from 'react';
import { hasError } from 'utils';
import {
  MockDetailedDataType,
  PokemonMoveDetails,
  PokemonTypeDetails,
  MockBaseDataType,
  PokemonDetails,
  CustomPokemon,
  ErrorsObject,
  CheckFields,
  PokemonData,
  PokemonMove,
  PokemonType,
  TextFields,
  Pokemon,
} from 'types';

export const visible = true;
export const invisible = false;
export const isLoading = false;

export const errorRef: RefObject<HTMLSpanElement> = createRef();
export const inputRef: RefObject<HTMLInputElement> = createRef();
export const selectRef: RefObject<HTMLSelectElement> = createRef();

export const links = ['Home', 'About', 'Form', 'Invalid'];
export const componentName = 'About';
export const location = '/about';
export const testingContent = 'Content';
export const testingTitle = 'Title';

export const TestingElement = () => <p>testing element</p>;
export const TestingErrorElement = () => <p>testing error</p>;

export const TestingTitleElement = () => <span>{testingTitle}</span>;
export const TestingContentElement = () => <div>{testingContent}</div>;
export const modalRef: RefObject<HTMLDivElement> = createRef();

export const fetchBaseMock = <T extends MockBaseDataType>(mockData: T) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve({ results: [mockData] }),
    })
  ) as jest.Mock;
};

export const fetchByParameterMock = <T extends MockDetailedDataType>(mockData: T) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve(mockData),
    })
  ) as jest.Mock;
};

export const fetchDetailsMock = <T,>(mockData: Array<T>) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve(mockData[0]),
    })
  ) as jest.Mock;
};

export const pokemonType: PokemonType = {
  name: 'typex',
  url: 'url',
};

export const pokemon: Pokemon = {
  name: 'pokemonix',
  url: 'url',
};

export const pokemonMove: PokemonMove = {
  name: 'movix',
  url: 'url',
};

export const typeModalContentProperties = [
  'Double damage from',
  'Double damage to',
  'Half damage from',
  'Half damage to',
  'No damage from',
  'No damage to',
  'Pokemons',
  'Moves',
] as const;

export const moveModalContentProperties = [
  'Learned by',
  'Priority',
  'Accuracy',
  'Power',
  'Type',
  'PP',
] as const;

export const pokemonModalContentProperties = [
  'Pokemon statistics',
  'Pokemon abilities',
  'Base experience',
  'Height',
  'Weight',
  'Types',
] as const;

export const pokemonData: PokemonData = {
  results: [{ name: 'testName', url: 'testURL' }],
  previous: null,
  next: null,
  count: 1,
};

export const testOnChange = (field: TextFields | CheckFields) => console.log(field);

export const testEmptyErrors: ErrorsObject = {
  birthday: null,
  consent: null,
  gender: null,
  name: null,
  type: null,
};

export const testErrors: ErrorsObject = {
  birthday: ErrorMessages.birthday,
  consent: ErrorMessages.consent,
  gender: ErrorMessages.gender,
  name: ErrorMessages.name,
  type: ErrorMessages.type,
};

export const customPokemon: CustomPokemon = {
  birthday: '01-01-1999',
  name: 'pokemonix',
  gender: 'male',
  avatar: null,
  type: 'fire',
  shiny: true,
};

export const detailedTypes: Array<PokemonTypeDetails> = [
  {
    damage_relations: {
      double_damage_from: [],
      double_damage_to: [{ name: 'toTest', url: 'urlTest' }],
      half_damage_from: [{ name: 'fromTest', url: 'urlTest' }],
      half_damage_to: [],
      no_damage_from: [],
      no_damage_to: [],
    },
    pokemon: [{ pokemon: { name: 'pokemonTest', url: 'urlTest' }, slot: 1 }],
    moves: [{ name: 'moveTest', url: 'urlTest' }],
    name: 'typeTest',
    id: 10,
  },
];

export const detailedMoves: Array<PokemonMoveDetails> = [
  {
    learned_by_pokemon: [{ name: 'pokomonTest', url: 'urlTest' }],
    type: { name: 'typeTest', url: 'urlTest' },
    name: 'moveTest',
    accuracy: 60,
    priority: 0,
    power: 50,
    id: 20,
    pp: 15,
  },
];

export const detailedPokemons: Array<PokemonDetails> = [
  {
    abilities: [{ ability: { name: 'ability' }, is_hidden: false, slot: 2 }],
    stats: [{ base_stat: 33, effort: 33, stat: { name: 'test' } }],
    sprites: { back_default: 'url', front_default: 'url' },
    types: [{ slot: 2, type: { name: 'typeTest' } }],
    name: 'Testing pokemon',
    base_experience: 12,
    height: 15,
    weight: 10,
    id: 34,
  },
  {
    abilities: [{ ability: { name: 'ability2' }, is_hidden: true, slot: 1 }],
    stats: [{ base_stat: 22, effort: 22, stat: { name: 'test' } }],
    sprites: { back_default: 'url2', front_default: 'url2' },
    types: [{ slot: 2, type: { name: 'typeTest2' } }],
    name: 'Testing pokemon2',
    base_experience: 16,
    height: 11,
    weight: 11,
    id: 35,
  },
];

type Store = Record<string, string>;

export const localStorageMock = (() => {
  let store: Store = {};
  return {
    getItem(key: string) {
      return store[key];
    },
    setItem(key: string, value: string) {
      store[key] = value.toString();
    },
    clear() {
      store = {};
    },
    removeItem(key: string) {
      delete store[key];
    },
  };
})();
