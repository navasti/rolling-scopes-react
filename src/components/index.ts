export { Loader } from './Loader';
export { Modal } from './Modal';
