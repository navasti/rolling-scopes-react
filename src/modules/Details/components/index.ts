export { MoveCardDetails } from './MoveCardDetails';
export { TypeCardDetails } from './TypeCardDetails';
export { PokemonCardDetails } from './PokemonCardDetails';
