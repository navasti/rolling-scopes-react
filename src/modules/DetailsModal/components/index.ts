export { TypeModalContent } from './TypeModalContent';
export { MoveModalContent } from './MoveModalContent';
export { PokemonModalContent } from './PokemonModalContent';
