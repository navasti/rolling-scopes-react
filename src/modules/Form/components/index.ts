export { BirthdayField } from './BirthdateField';
export { ErrorMessage } from './ErrorMessage';
export { ConsentField } from './ConsentField';
export { AvatarField } from './AvatarField';
export { GenderField } from './GenderField';
export { ShinyField } from './ShinyField';
export { TypeField } from './TypeField';
export { NameField } from './NameField';
export { FormCard } from './FormCard';
