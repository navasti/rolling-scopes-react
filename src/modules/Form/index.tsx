import React, { createRef, FormEvent, RefObject } from 'react';
import { CheckFields, CustomPokemon, ErrorsObject, TextFields } from 'types';
import { Layout } from 'modules';
import * as S from './styled';
import { uuid } from 'utils';
import {
  BirthdayField,
  ErrorMessage,
  ConsentField,
  GenderField,
  AvatarField,
  ShinyField,
  TypeField,
  NameField,
  FormCard,
} from 'modules/Form/components';
import { FIELDS_VALIDATION_BY_NAME, FEMALE, MALE, ErrorMessages } from 'appConstants';

type Props = {
  componentName: string;
  location: string;
};

type State = {
  pokemons: Array<CustomPokemon & { id: string }>;
  initialEnter: boolean;
  errors: ErrorsObject;
};

export class Form extends React.Component<Props, State> {
  femaleInputRef: RefObject<HTMLInputElement>;
  successMessage: RefObject<HTMLSpanElement>;
  birthdayField: RefObject<HTMLInputElement>;
  isShinyField: RefObject<HTMLInputElement>;
  consentField: RefObject<HTMLInputElement>;
  maleInputRef: RefObject<HTMLInputElement>;
  avatarField: RefObject<HTMLInputElement>;
  typeField: RefObject<HTMLSelectElement>;
  nameField: RefObject<HTMLInputElement>;

  constructor(props: Props) {
    super(props);
    this.state = {
      pokemons: [],
      initialEnter: true,
      errors: {
        birthday: null,
        consent: null,
        gender: null,
        name: null,
        type: null,
      },
    };
    this.successMessage = createRef();
    this.femaleInputRef = createRef();
    this.birthdayField = createRef();
    this.consentField = createRef();
    this.isShinyField = createRef();
    this.maleInputRef = createRef();
    this.avatarField = createRef();
    this.typeField = createRef();
    this.nameField = createRef();
  }

  removeError = (field: TextFields | CheckFields) => {
    if (this.state.initialEnter) this.setState({ initialEnter: false });
    const errors = { ...this.state.errors };
    errors[field] = null;
    this.setState({ errors });
  };

  validate = (fields: Array<React.RefObject<HTMLInputElement | HTMLSelectElement>>) => {
    const errors = { ...this.state.errors };
    const textFields = FIELDS_VALIDATION_BY_NAME.TEXT.map((textField) =>
      fields.find((field) => field.current?.name === textField)
    );
    const checkFields = FIELDS_VALIDATION_BY_NAME.CHECK.map((checkField) =>
      fields.filter((field) => field.current?.name === checkField)
    );
    checkFields.forEach((checkField) => {
      const valid = checkField.some((field) => (field.current as HTMLInputElement)?.checked);
      const name = checkField[0]?.current?.name
        ? (checkField[0].current.name as CheckFields)
        : null;
      if (!valid && name) errors[name] = ErrorMessages[name];
    });
    textFields.forEach((textField) => {
      const name = textField?.current?.name ? (textField.current.name as TextFields) : null;
      const value = textField?.current?.value;
      const valid = value && value.trim().length >= 2;
      if (!valid && name) errors[name] = ErrorMessages[name];
    });
    return errors;
  };

  prepareCustomPokemon = () => {
    const shiny = !!this.isShinyField.current?.checked;
    const female = this.femaleInputRef?.current?.checked;
    const file = this.avatarField.current?.files?.[0] || null;
    const customPokemon: CustomPokemon & { id: string } = {
      birthday: this.birthdayField?.current?.value || '',
      type: this.typeField?.current?.value || '',
      name: this.nameField?.current?.value || '',
      gender: female ? FEMALE : MALE,
      avatar: file,
      id: uuid(),
      shiny,
    };
    return customPokemon;
  };

  handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const fields = [
      this.nameField,
      this.typeField,
      this.avatarField,
      this.isShinyField,
      this.consentField,
      this.birthdayField,
      this.maleInputRef,
      this.femaleInputRef,
    ];
    const errors = this.validate(fields);
    if (Object.values(errors).some((error) => error != null)) {
      this.setState({ errors });
    } else {
      const customPokemon = this.prepareCustomPokemon();
      this.setState({
        pokemons: [...this.state.pokemons, customPokemon],
      });
      fields.forEach((field) => {
        const current = field.current;
        if (current) {
          if (current.type === 'radio' || current.type === 'checkbox') {
            (current as HTMLInputElement).checked = false;
          } else current.value = '';
        }
      });
    }
  };

  render() {
    const { location, componentName } = this.props;
    return (
      <Layout location={location} componentName={componentName}>
        <S.CommonView>
          <S.FormHeading>Create custom pokemon!</S.FormHeading>
          <S.Form onSubmit={this.handleSubmit}>
            <NameField
              onChange={this.removeError}
              errors={this.state.errors}
              ref={this.nameField}
            />
            <TypeField
              onChange={this.removeError}
              errors={this.state.errors}
              ref={this.typeField}
            />
            <S.GenderWrapper>
              <S.RadioWrapper>
                *Gender
                <S.RadioFields>
                  <GenderField
                    name={FEMALE}
                    ref={this.femaleInputRef}
                    onChange={this.removeError}
                  />
                  <GenderField name={MALE} ref={this.maleInputRef} onChange={this.removeError} />
                </S.RadioFields>
              </S.RadioWrapper>
              <ErrorMessage
                visible={!!this.state.errors.gender}
                message={this.state.errors.gender}
              />
            </S.GenderWrapper>
            <BirthdayField
              ref={this.birthdayField}
              errors={this.state.errors}
              onChange={this.removeError}
            />
            <AvatarField ref={this.avatarField} />
            <ShinyField ref={this.isShinyField} />
            <ConsentField
              ref={this.consentField}
              errors={this.state.errors}
              onChange={this.removeError}
            />
            <S.SubmitButton
              disabled={
                this.state.initialEnter ||
                Object.values(this.state.errors).some((error) => error != null)
              }
              type="submit"
            >
              Submit
            </S.SubmitButton>
            <S.SuccessMessage ref={this.successMessage}>Pokemon added successfuly</S.SuccessMessage>
          </S.Form>
          <hr />
          <S.CardsGrid>
            {this.state.pokemons.map((pokemon) => (
              <FormCard key={pokemon.id} customPokemon={pokemon} />
            ))}
          </S.CardsGrid>
        </S.CommonView>
      </Layout>
    );
  }
}
