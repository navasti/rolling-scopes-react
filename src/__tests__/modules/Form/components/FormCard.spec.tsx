import { screen, render } from '@testing-library/react';
import { FormCard } from 'modules/Form/components';
import { capitalize } from 'utils';
import { customPokemon } from '__mocks__/data';

describe('FormCard', () => {
  it('should be rendered and match snapshot', () => {
    const { container } = render(<FormCard customPokemon={customPokemon} />);
    expect(screen.getByText(customPokemon.birthday)).toBeVisible();
    expect(screen.getByText(capitalize(customPokemon.name))).toBeVisible();
    expect(screen.getByText(capitalize(customPokemon.gender))).toBeVisible();
    expect(screen.getByText(customPokemon.shiny ? 'Yes' : 'No')).toBeVisible();
    expect(container).toMatchSnapshot();
  });
  it('image should be visible and src set as questionmark.png', () => {
    const { container } = render(<FormCard customPokemon={customPokemon} />);
    const img = container.querySelector('img') as HTMLImageElement;
    expect(img).toHaveAttribute('src', 'questionmark.png');
    expect(img).toBeVisible();
  });
});
