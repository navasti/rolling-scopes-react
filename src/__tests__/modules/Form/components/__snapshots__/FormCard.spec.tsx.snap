// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormCard should be rendered and match snapshot 1`] = `
<div>
  <div
    class="sc-ftvSup gMTpqP"
  >
    <div
      class="sc-papXJ jiIdkW"
    >
      <img
        src="questionmark.png"
      />
    </div>
    <div
      class="sc-jqUVSM kQjLTY"
    >
      <p>
        Name
        : 
        <span>
          Charmander
        </span>
      </p>
      <p>
        Gender
        : 
        <span>
          Male
        </span>
      </p>
      <p>
        Main type
        : 
        <span>
          Fire
        </span>
      </p>
      <p>
        Birthday
        : 
        <span>
          01-01-1999
        </span>
      </p>
      <p>
        Shiny
        : 
        <span>
          Yes
        </span>
      </p>
    </div>
  </div>
</div>
`;
